feat(graphql,auth,#1026): Pass operation name to authorizer

mwoelk opened the pull request:
Hey Doug,
here's a pull request that adds the primary feature requested in #1026.
I'm looking forward to your feedback.

**Contents**
- Updated `Authorizer` interface to also receive an `operationName` (optional Argument to keep backwards compatibility)
- Updated `DefaultCrudAuthorizer` and `AuthorizerOptions` to also pass along the `operationName` argument.
- Updated `AuthorizerFilter`, `RelationAuthorizerFilter` and `ModifyRelationAuthorizerFilter` to pass the name of the annotated method as `operationName` to the `Authorizer`. This name can also be overriden by a new optional argument `operationName` that was added to those decorators.
- Used this override on `queryMany` of the `ReadResolver` to change the passed name to `query` to better align with the names of the query service.
- Updated the auth-example to show new possibilities using the `nestjs-query-3` - user. Modified the example so that this user can read and aggregate all todos of all users but only update/delete his own todos
- Added two new unit tests that check if the `operationName` parameter gets passed correctly
- Extended documentation on authorization  based on operation names

**Unimplemented parts of the feature request**
- Does not pass the user input/args to the authorizer because this would require deeper changes and might overcomplicate things. Might be better to perform such checks using custom validators instead.
- Authorization on `create*` methods. This should be the scope of a new issue/PR. Main use case would be to throw an Unauthorized here if a user isn't authorized. The filters returned by the authorizers may not be needed for creation. (But maybe could be checked against the input but this would be more related to validation again).

doug-martin commented:
@mwoelk thank you for the PR! 

This makes sense to me, the only thing I question is if the we should be passing the name directly or instead create a new type (maybe `AuthorizerContext`) that has a field `operationName`.I think this makes for a more flexible implementation that won't be tied to passing additional params to make the change passive, instead we can just add new fields to the `AuthorizerContext` as use cases are discovered.

mwoelk commented:
@doug-martin sounds great. Thought about it in the beginning but wasn't sure of any additional parameters but it's definitely the better approach to keep it flexible. I'm going to update the PR today.

mwoelk commented:
@doug-martin Here's a summary of the changes made in the three added commits:

1. As you proposed, I've refactored the argument into a new type (I've went with `AuthorizationContext` because `AuthorizerContext` is already defined for the `AuthorizerInterceptor` and `AuthorizationContext` also seems more fitting because it's varies for every single authorization). (refactored and updated docs in 81ea148)
2. I didn't like the way one had to query for each method name if one were to implement e.g. a different logic for read/write methods so I added a few more fields to the `AuthorizationContext` in 13a4d73. More explanation below.
3. With the changes made in 2, it only seemed logical to me to add the `@AuthorizationFilter` in 11be43e to the create methods as well to at least give the user the possibility to throw an `UnauthorizedException` if we don't want someone to create an entity. Right now those filters are unused but it would be illogical to have all the nice `Authorizer` logic setup and having to write additional logic only to authorize the create methods (e.g. using custom guards / decorators). 

## AuthorizationContext structure
```ts
interface AuthorizationContext {
  /** The name of the method that uses the @AuthorizeFilter decorator */
  operationName: string;

  /** The group this operation belongs to */
  operationGroup: 'read' | 'aggregate' | 'create' | 'update' | 'delete';

  /** If the operation does not modify any entities */
  readonly: boolean;

  /** If the operation can affect multiple entities */
  many: boolean;
}
```

This way in a custom `Authorizer` we can easily check different requirements instead of having to repeat a lot of conditions. ([see auth example](https://github.com/doug-martin/nestjs-query/pull/1043/files/d49db8bb76ca05419668cde7f6f49be6696bf1a0..13a4d73953ed7200ec4996434c08edaff89bf250?file-filters%5B%5D=.json&file-filters%5B%5D=.ts#diff-cd0391ed319e31e4c661f3570b403fc247d1b3b3538ce17fce5647ee52802d00))

e.g. we can perform a simple
```ts
if(authContext.readonly) {
...
}
```
instead of comparing the operation name to all possible read operations such as query, aggregate, find, etc.

## AuthorizationContext inference
This context is either passed as an argument to the `@AuthorizationFilter` or inferred by the methodName / name passed to the `@AuthorizationFilter`. Inference is done in a way that fits the naming conventions of the auto generated resolvers:
- if the method name starts with `query`, `find` or `aggregate`, `readonly` is set to `true`
- if the name starts with `query` or `aggregate` or ends with `many`, `many` is set to `true`
- if the method name starts with `create`, `delete`, `update`, `aggregate`, the `operationGroup` is set accordingly. Otherwise it's set to `read`

## AuthorizationFilter
The authorization filter can be used in three ways:
- Without arguments `@AuthorizationFilter()`, infers the context from the name of the decorated method
- With a name as argument `@AuthorizationFilter('queryCompletedTodoItems')`, infers the context using the provided name
- With an `AuthorizationContext` as argument:
  ```ts
  @AuthorizerFilter({
    operationName: 'completedTodoItems',
    operationGroup: 'read',
    readonly: true,
    many: true
  })
  ```

doug-martin commented on packages/query-graphql/src/resolvers/read.resolver.ts:
```diff
@@ -86,7 +86,7 @@ export const Readable = <DTO, ReadOpts extends ReadResolverOpts<DTO>, QS extends
     async queryMany(
       @HookArgs() query: QA,
-      @AuthorizerFilter() authorizeFilter?: Filter<DTO>,
+      @AuthorizerFilter('query') authorizeFilter?: Filter<DTO>,
```
Why does `query` need to be specified, is it not gathered from the `propertyKey`?

mwoelk replied:
I just added query to change the methodName in the context to query to closer align with the names known form the docs. This change was from the beginning when only the method name was passed. I guess with the extended context we might just leave it empty to be infered as `queryMany`. 
@doug-martin What do you think?

mwoelk replied:
Changed to use original methodName (`queryMany`) in b6c7ce9

doug-martin commented on packages/query-graphql/src/decorators/authorize-filter.decorator.ts:
```diff
@@ -1,47 +1,117 @@
-  return createParamDecorator((data: unknown, executionContext: ExecutionContext) =>
-    getAuthorizerFilter(getContext<AuthorizerContext<DTO>>(executionContext)),
-  )();
+function getAuthorizationContext(operationNameOrContext: string | AuthorizationContext): AuthorizationContext {
```
While I like this, I'm also torn on it. I'm concerned about it relying on convention without any checks to ensure that the methodName conforms to a known convention. Without the check I think it could be easy for an implementor to add a new method that does not conform to the convention and their authorizer will not work.

I think there are a couple of paths forward.

1. Require the context to be set explicitly, this would allow for easier internal refactors when looking for usages of the `AuthorizationContext`. 
2. Throw an error if we can not determine the operation group/type. Looking at how its initialized I think throwing an error would be ok in this instance as it occurs when the decorator is run.

mwoelk replied:
Hmm... what about backwards compatibility or if someone just wants to use `Authorizers` the basic old way?
If a user doesn't want to use the context he shouldn't be forced to pass one. In this case we could just pass `undefined` as context and keep it optional in the `Authorizer`.  We might as well log a warning in the decorator. This way it will be logged only once at startup to inform the user.

The perfect way would be to check if the `Authorizer` is using the second argument of the `authorize` function but I don't think Typescript reflection offers this possibility (at least not without using additional decorators)

mwoelk replied:
Maybe we should just leave out the whole inference part? Just realised that the [relation resolvers](https://github.com/doug-martin/nestjs-query/blob/da60b9340879258df6b5294c3b43fff1f6bcc177/packages/query-graphql/src/resolvers/relations/update-relations.resolver.ts#L33) don't use the `Many`-prefix but the singular/plural forms of the relation DTO. So I guess the future-proof method would be to refactor the resolvers to explicitly pass the context which would also help developers looking up the source code.

mwoelk replied:
Removed the inference part and updated resolvers in b6c7ce9.

@doug-martin : Still not sure about passing undefined / throwing error / logging warning...

mwoelk replied:
Maybe instead of returning undefined we could return a proxy that throws an error as soon as one of the properties is accessed. Sadly this error wouldn't be thrown upfront but just-in-time when the user wants to use context-based authorization and hasn't provided any context to his custom method.

mwoelk replied:
You can check out the proxy-solution in this [branch](https://github.com/mwoelk/nestjs-query/tree/authorizer-proxy)

doug-martin commented on packages/query-graphql/src/decorators/authorize-filter.decorator.ts:
```diff
@@ -1,47 +1,117 @@
+
+  const lcMethodName = operationNameOrContext.toLowerCase();
+
+  const isCreate = lcMethodName.startsWith('create');
```
Related to my above comment, move these magic strings to either an enum or constants.

mwoelk replied:
Obsolete in b6c7ce9

doug-martin reviewed (changes requested):
Wow! Great job diving in and figuring this out so quickly. I had a few comments let me know what you think. 

Thank you again for your contribution, this is awesome!

mwoelk commented on packages/query-graphql/src/auth/authorizer.ts:
```diff
@@ -1,9 +1,29 @@
 import { Filter } from '@nestjs-query/core';
 
+export type AuthorizationOperationGroup = 'read' | 'aggregate' | 'create' | 'update' | 'delete';
```
Fixed in b6c7ce9

mwoelk commented on packages/query-graphql/src/auth/authorizer.ts:
```diff
@@ -1,9 +1,29 @@
 export interface Authorizer<DTO> {
   // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types,@typescript-eslint/no-explicit-any
-  authorize(context: any): Promise<Filter<DTO>>;
+  authorize(context: any, authorizerContext?: AuthorizationContext): Promise<Filter<DTO>>;
```
Related to comments on packages/query-graphql/src/decorators/authorize-filter.decorator.ts below

mwoelk replied:
And it wouldn't be backwards compatible if someone is using the injected `Authorizer` directly:
```ts
@InjectAuthorizer(TodoItemDTO) readonly authorizer: Authorizer<TodoItemDTO>
```

doug-martin replied:
@mwoelk good point, since we're on a v0.x.x.` release still I'm ok doing a minor version bump to introduce a breaking change.

mwoelk replied:
Ok. Already thrown out the optional stuff in the branch mentioned below.

doug-martin commented:
@mwoelk the proxy branch looks really good. The two things that stuck out to me after a first pass is the `many` option in the [`DeleteResolver#deleteMany`](https://github.com/doug-martin/nestjs-query/compare/master...mwoelk:authorizer-proxy#diff-680b5b5144956a861452260140865c9b8ac03b993475bb465666a3e0553647a2R132) and adding an example of using the `authorizationContext`.

Once you make those changes I'll do one final pass merge and release! Thanks again!

mwoelk commented:
@doug-martin Good catch with the wrong attribute on the deleteMany context (fixed in 9306525). Regarding the examples what kind of example do you mean? I added a whole section to the [docs](https://github.com/doug-martin/nestjs-query/pull/1043/files#diff-ea814d7f86f41ad91d38899bf99efd476553859d71d8d93207e18378ff373c43R402-R452) and added some example usage to the auth-example in the [SubTaskAuthorizer](https://github.com/doug-martin/nestjs-query/pull/1043/files?file-filters%5B%5D=.json&file-filters%5B%5D=.ts#diff-cd0391ed319e31e4c661f3570b403fc247d1b3b3538ce17fce5647ee52802d00R7-R10) and in the [TodoItemDTO](https://github.com/doug-martin/nestjs-query/pull/1043/files?file-filters%5B%5D=.json&file-filters%5B%5D=.ts#diff-f9966919586c3ec391a24e172df623a7e081701b2e4221a476c5be075bec5e8cR18-R31)

Should I create a whole new example project based on the auth example that covers different use cases of the authContext?
